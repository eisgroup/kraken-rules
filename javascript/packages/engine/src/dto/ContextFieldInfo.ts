/*
 *  Copyright 2020 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Contexts } from "kraken-model";
import { DataContext } from "../engine/contexts/data/DataContext";
import { ErrorCode, KrakenRuntimeError } from "../error/KrakenRuntimeError";

export class ContextFieldInfo {
    readonly contextId: string;
    readonly contextName: string;
    readonly fieldName: string;
    readonly fieldPath: string;
    readonly fieldType: string;
    constructor(dataContext: DataContext, fieldName: string) {
        let projection: Contexts.ContextField;
        if (dataContext.definitionProjection && dataContext.definitionProjection[fieldName]) {
            projection = dataContext.definitionProjection[fieldName];
        } else {
            throw new KrakenRuntimeError(
                ErrorCode.INCORRECT_MODEL_TREE,
                `Fields projection for a DataContext instance lacks of field '${fieldName}'`
            );
        }
        this.contextId = dataContext.contextId;
        this.contextName = dataContext.contextName;
        this.fieldName = fieldName;
        this.fieldPath = projection.fieldPath;
        this.fieldType = projection.fieldType;
    }
}
