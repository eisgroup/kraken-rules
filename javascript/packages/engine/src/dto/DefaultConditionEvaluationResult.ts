/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { ConditionEvaluation, ConditionEvaluationResult, ExpressionEvaluationResult } from 'kraken-engine-api'

export class DefaultConditionEvaluationResult implements ConditionEvaluationResult {
    private constructor(
        public readonly conditionEvaluation: ConditionEvaluation,
        public readonly error: ExpressionEvaluationResult.ErrorResult | undefined,
    ) {}

    static fromError = (error: ExpressionEvaluationResult.ErrorResult) => {
        return new DefaultConditionEvaluationResult(ConditionEvaluation.ERROR, error)
    }

    static of = (evaluation: ConditionEvaluation) => new DefaultConditionEvaluationResult(evaluation, undefined)
}

export const conditionEvaluationTypeChecker = {
    isApplicable(result: ConditionEvaluationResult): boolean {
        return result.conditionEvaluation === ConditionEvaluation.APPLICABLE
    },
}
