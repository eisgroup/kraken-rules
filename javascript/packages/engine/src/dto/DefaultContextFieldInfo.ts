/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Contexts } from 'kraken-model'
import { DataContext } from '../engine/contexts/data/DataContext'
import { CONTEXT_MODEL_TREE_MISSING_FIELD, KrakenRuntimeError, SystemMessageBuilder } from '../error/KrakenRuntimeError'
import { ContextFieldInfo } from 'kraken-engine-api'

export class DefaultContextFieldInfo implements ContextFieldInfo {
    readonly contextId: string
    readonly contextName: string
    readonly fieldName: string
    readonly fieldPath: string
    readonly fieldType: string

    constructor(dataContext: DataContext, fieldName: string) {
        let contextField: Contexts.ContextField
        if (dataContext.contextDefinition.fields?.[fieldName]) {
            contextField = dataContext.contextDefinition.fields[fieldName]
        } else {
            const m = new SystemMessageBuilder(CONTEXT_MODEL_TREE_MISSING_FIELD).parameters(fieldName).build()
            throw new KrakenRuntimeError(m)
        }
        this.contextId = dataContext.contextId
        this.contextName = dataContext.contextName
        this.fieldName = fieldName
        this.fieldPath = contextField.fieldPath
        this.fieldType = contextField.fieldType
    }
}
