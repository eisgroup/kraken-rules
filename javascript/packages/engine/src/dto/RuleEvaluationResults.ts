/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { PayloadResult, ValidationPayloadResult } from "../engine/results/PayloadResult";
import { ConditionEvaluationResult } from "./ConditionEvaluationResult";
import { RuleInfo } from "../engine/results/RuleInfo";
import { RuleOverride } from "..";

export namespace RuleEvaluationResults {

    export enum Kind {
        REGULAR = 1,
        VALIDATION = 2,
        NOT_APPLICABLE = 4
    }

    interface BaseValidationResult {
        kind: Kind;
        ruleInfo: RuleInfo;
        conditionEvaluationResult: ConditionEvaluationResult;
    }

    export interface NotApplicableRuleEvaluationResult extends BaseValidationResult {
        kind: Kind.NOT_APPLICABLE;
    }

    export interface RegularRuleEvaluationResult extends BaseValidationResult {
        kind: Kind.REGULAR;
        payloadResult: PayloadResult;
    }

    export interface ValidationRuleEvaluationResult extends BaseValidationResult {
        kind: Kind.VALIDATION;
        payloadResult: ValidationPayloadResult;
        overrideInfo: OverrideInfo;
    }

    export type RuleEvaluationResult = ValidationRuleEvaluationResult
        | RegularRuleEvaluationResult
        | NotApplicableRuleEvaluationResult;

    export type ApplicableRuleEvaluationResult = ValidationRuleEvaluationResult
        | RegularRuleEvaluationResult;

    export function isValidation(r: RuleEvaluationResult): r is ValidationRuleEvaluationResult {
        return Boolean(r.kind & Kind.VALIDATION);
    }

    export function isNotApplicable(r: RuleEvaluationResult): r is NotApplicableRuleEvaluationResult {
        return Boolean(r.kind & Kind.NOT_APPLICABLE);
    }

    export interface OverridePresentInfo {
        /**
         * when rule is failed and rule is overridable
         */
        overrideApplicable: true;
        overridable: true;
        overrideGroup?: string;
        overrideContext: RuleOverride.OverridableRuleContextInfo;
    }

    export interface OverrideNotPresentInfo {
        overrideApplicable: false;
        overridable: boolean;
        overrideGroup?: string;
    }

    export type OverrideInfo = OverridePresentInfo | OverrideNotPresentInfo;

    export const overrideInfoTypeChecker = {
        isOverrideApplicable(p: OverrideInfo): p is OverridePresentInfo {
            return p.overrideApplicable;
        }
    };
}
