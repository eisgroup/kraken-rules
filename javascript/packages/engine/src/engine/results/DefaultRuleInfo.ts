/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Payloads, Rule } from 'kraken-model'
import { RuleInfo } from 'kraken-engine-api'

export class DefaultRuleInfo implements RuleInfo {
    readonly ruleName: string
    readonly context: string
    readonly targetPath: string
    readonly payloadtype: Payloads.PayloadType

    constructor(rule: Rule) {
        this.context = rule.context
        this.ruleName = rule.name
        this.targetPath = rule.targetPath
        this.payloadtype = rule.payload.type
    }
}
