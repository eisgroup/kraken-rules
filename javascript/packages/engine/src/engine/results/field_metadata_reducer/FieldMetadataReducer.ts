/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import {
    RuleEvaluationResults,
    EntryPointResult,
    RuleOverride,
    ContextFieldInfo,
    ErrorMessage,
    FieldMetadata,
    FieldMetadataResult,
} from 'kraken-engine-api'

import { EntryPointReducer } from '../Reducer'
import { Localization } from '../Localization'
import ValidationRuleEvaluationResult = RuleEvaluationResults.ValidationRuleEvaluationResult
import { payloadResultTypeChecker } from '../PayloadResultTypeChecker'

/**
 * Processes rule evaluation information from {@link EntryPointResult}, reducing it to list of
 * metadata entries for each distinct field
 */
export class FieldMetadataReducer implements EntryPointReducer<Record<string, FieldMetadata>> {
    readonly #isRuleOverridden: RuleOverride.IsRuleOverridden
    readonly #errorMessageFrom: (rer: ValidationRuleEvaluationResult, contextInfo: ContextFieldInfo) => ErrorMessage

    constructor(
        isRuleOverridden?: RuleOverride.IsRuleOverridden,
        validationMessageProvider?: Localization.ValidationMessageProvider,
    ) {
        this.#isRuleOverridden = isRuleOverridden ? isRuleOverridden : () => false

        if (validationMessageProvider) {
            this.#errorMessageFrom = rer => Localization.resolveMessage(rer.payloadResult, validationMessageProvider)
        } else {
            this.#errorMessageFrom = rer =>
                Localization.resolveMessage(rer.payloadResult, Localization.defaultValidationMessages)
        }
    }

    /**
     * @override
     */
    reduce(results: EntryPointResult): Record<string, FieldMetadata> {
        const fieldMetadataResults: Record<string, FieldMetadata> = {}
        const fieldResults = results.getFieldResults()
        for (const id in fieldResults) {
            if (Object.prototype.hasOwnProperty.call(fieldResults, id)) {
                const fieldResult = fieldResults[id]
                let isDisabled
                let isHidden
                const validationResults: FieldMetadataResult[] = []
                for (const ruleResult of fieldResult.ruleResults) {
                    if (!RuleEvaluationResults.isNotApplicable(ruleResult)) {
                        const payloadResult = ruleResult.payloadResult
                        // resolve accessibility
                        if (!isDisabled && payloadResultTypeChecker.isAccessibility(payloadResult)) {
                            isDisabled = !payloadResult.accessible
                        }
                        // resolve visibility
                        if (!isHidden && payloadResultTypeChecker.isVisibility(payloadResult)) {
                            isHidden = !payloadResult.visible
                        }
                        // resolve validation
                        if (RuleEvaluationResults.isValidation(ruleResult)) {
                            const error = this.#errorMessageFrom(ruleResult, fieldResult.contextFieldInfo)
                            const isOverridden =
                                RuleEvaluationResults.overrideInfoTypeChecker.isOverrideApplicable(
                                    ruleResult.overrideInfo,
                                ) && this.#isRuleOverridden(ruleResult.overrideInfo.overrideContext)
                            validationResults.push({
                                rawTemplateVariables: ruleResult.payloadResult.message?.rawTemplateVariables ?? [],
                                ruleName: ruleResult.ruleInfo.ruleName,
                                errorCode: error.errorCode,
                                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                                errorMessage: error.errorMessage!,
                                templateVariables: error.templateVariables,
                                payloadResult,
                                severity: ruleResult.payloadResult.validationSeverity,
                                isFailed:
                                    ruleResult.payloadResult.success !== undefined
                                        ? !ruleResult.payloadResult.success
                                        : undefined,
                                isOverridable: ruleResult.overrideInfo.overridable,
                                isOverridden,
                            })
                        }
                    }
                }

                // register results
                fieldMetadataResults[id] = {
                    id,
                    info: {
                        getContextInstanceId: () => fieldResult.contextFieldInfo.contextId,
                        getContextName: () => fieldResult.contextFieldInfo.contextName,
                    },
                    resolvedTargetPath: fieldResult.contextFieldInfo.fieldPath,
                    isDisabled,
                    isHidden,
                    ruleResults: validationResults,
                    fieldType: fieldResult.contextFieldInfo.fieldType,
                }
            }
        }
        return fieldMetadataResults
    }
}
