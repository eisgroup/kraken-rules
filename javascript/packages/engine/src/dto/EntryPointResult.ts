/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { FieldEvaluationResult } from "./FieldEvaluationResult";
import { RuleEvaluationResults } from "./RuleEvaluationResults";
import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult;

import { Reducer } from "declarative-js";
import flat = Reducer.flat;
import { ConditionEvaluationResult } from "./ConditionEvaluationResult";

/**
 * Contains evaluation results from specific entry point
 */
export class EntryPointResult {
    private readonly results: Record<string, FieldEvaluationResult>;
    readonly evaluationTimestamp: Date;
    constructor(results: Record<string, FieldEvaluationResult>) {
        this.results = results;
        this.getApplicableResults.bind(this);
        this.evaluationTimestamp = new Date();
    }
    /**
     * Flatten contents into single list of {@link RuleEvaluationResult}s
     */
    getAllRuleResults(): RuleEvaluationResult[] {
        return Object.values(this.results)
            .map(x => x.ruleResults)
            .reduce(flat, []);
    }
    getFieldResults(): Record<string, FieldEvaluationResult> {
        return this.results;
    }

    getApplicableResults(): RuleEvaluationResult[] {
        return this.getAllRuleResults()
            .filter(r => ConditionEvaluationResult.isApplicable(r.conditionEvaluationResult));
    }

    static empty(): EntryPointResult {
        return new EntryPointResult({});
    }
}
