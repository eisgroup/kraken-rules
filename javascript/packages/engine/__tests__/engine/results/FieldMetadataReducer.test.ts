/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
import { Reducer } from "declarative-js";

import { EntryPointResult } from "../../../src/dto/EntryPointResult";
import { FieldEvaluationResultImpl } from "../../../src/dto/FieldEvaluationResult";
import { mock } from "../../mock";
import { RuleEvaluationResults } from "../../../src/dto/RuleEvaluationResults";

import {
    AccessibilityPayloadResult,
    DefaultValuePayloadResult,
    VisibilityPayloadResult,
    PayloadResultType,
    payloadResultCreator,
    payloadResultTypeChecker
} from "../../../src/engine/results/PayloadResult";
import { FieldMetadataReducer } from "../../../src/engine/results/field_metadata_reducer/FieldMetadataReducer";
import { FieldMetadata } from "../../../src/engine/results/field_metadata_reducer/FieldMetadata";
import { ValueChangedEvent } from "../../../src/engine/results/Events";
import { ConditionEvaluation, ConditionEvaluationResult } from "../../../src/dto/ConditionEvaluationResult";
import { PayloadBuilder, RulesBuilder } from "kraken-model-builder";
import { FieldMetadataMocks } from "./mocks/FieldMetadataReducer.mocks";
import { RuleOverride } from "../../../src/engine/results/RuleOverride";
import { RuleInfo } from "../../../src/engine/results/RuleInfo";
import { ContextFieldInfo } from "../../../src/dto/ContextFieldInfo";
import { Payloads } from "kraken-model";

const { Policy } = mock.modelTreeJson.contexts;
let reducedResults: Record<string, FieldMetadata>;
let entryPointResults: EntryPointResult;
const rule = (targetPath: string) => RulesBuilder.create()
    .setContext("mock")
    .setTargetPath(targetPath)
    .setPayload(PayloadBuilder.lengthLimit().overridableLimit(0))
    .setName("mock")
    .build();
beforeEach(() => {
    const results = {} as Record<string, FieldEvaluationResultImpl>;
    results["present1"] = new FieldEvaluationResultImpl(
        new ContextFieldInfo(
            mock.data.dataContextCustom({ state: "AZ" }),
            Policy.fields.state.name
        ),
        [
            {
                kind: RuleEvaluationResults.Kind.REGULAR,
                conditionEvaluationResult: ConditionEvaluationResult.of(ConditionEvaluation.APPLICABLE),
                payloadResult: {
                    type: PayloadResultType.ACCESSIBILITY,
                    accessible: false
                } as AccessibilityPayloadResult,
                ruleInfo: new RuleInfo(rule(Policy.fields.state.name))
            }
        ]
    );
    results["present2"] = new FieldEvaluationResultImpl(
        new ContextFieldInfo(
            mock.data.dataContextCustom({ termDetails: { termCd: "hu" } }),
            Policy.fields.termCd.name
        ),
        [
            {
                kind: RuleEvaluationResults.Kind.REGULAR,
                conditionEvaluationResult: ConditionEvaluationResult.of(ConditionEvaluation.APPLICABLE),
                payloadResult: {
                    type: PayloadResultType.VISIBILITY,
                    visible: false
                } as VisibilityPayloadResult,
                ruleInfo: new RuleInfo(rule(Policy.fields.termCd.name))
            }
        ]
    );
    const valueChangedEvent = new ValueChangedEvent("a.b", "name2", "1", "new", "old");
    results["2"] = new FieldEvaluationResultImpl(
        new ContextFieldInfo(
            mock.data.dataContextCustom({ transactionDetails: { txType: "j" } }),
            Policy.fields.txType.name
        ),
        [
            {
                kind: RuleEvaluationResults.Kind.REGULAR,
                conditionEvaluationResult: ConditionEvaluationResult.of(ConditionEvaluation.APPLICABLE),
                payloadResult: {
                    type: PayloadResultType.DEFAULT,
                    events: [valueChangedEvent]
                } as DefaultValuePayloadResult,
                ruleInfo: new RuleInfo(rule(Policy.fields.txType.name))
            }
        ]
    );
    results["4"] = new FieldEvaluationResultImpl(
        new ContextFieldInfo(
            mock.data.dataContextCustom({ termDetails: { termNo: 6 } }),
            Policy.fields.termNo.name
        ),
        [
            {
                kind: RuleEvaluationResults.Kind.REGULAR,
                conditionEvaluationResult: ConditionEvaluationResult.of(ConditionEvaluation.APPLICABLE),
                payloadResult: {
                    type: PayloadResultType.DEFAULT,
                    events: [valueChangedEvent]
                } as DefaultValuePayloadResult,
                ruleInfo: new RuleInfo(rule(Policy.fields.termNo.name))
            }
        ]
    );
    results["3"] = new FieldEvaluationResultImpl(
        new ContextFieldInfo(
            mock.data.dataContextCustom({ accessTrackInfo: { updatedBy: "override" } }),
            Policy.fields.updatedBy.name
        ),
        [
            {
                kind: RuleEvaluationResults.Kind.VALIDATION,
                conditionEvaluationResult: ConditionEvaluationResult.of(ConditionEvaluation.APPLICABLE),
                payloadResult: payloadResultCreator.assertion(
                    PayloadBuilder.asserts().that(""),
                    false,
                    []
                ),
                ruleInfo: new RuleInfo(rule(Policy.fields.termNo.name)),
                overrideInfo: {
                    overridable: true,
                    overrideApplicable: true,
                    overrideContext: {
                        contextAttributeValue: "override",
                        contextId: "",
                        contextName: "",
                        overrideDependencies: {},
                        rootContextId: "",
                        rule: new RuleInfo(rule(Policy.fields.updatedBy.name)),
                        ruleEvaluationTimeStamp: new Date()
                    }
                }
            }
        ]
    );
    entryPointResults = new EntryPointResult(results);
    reducedResults = new FieldMetadataReducer().reduce(entryPointResults);
});
describe("FieldMetadataReducer", () => {
    it("should contain 2 field metadatas", () => {
        expect(Object.values(reducedResults)).toHaveLength(5);
        expect(reducedResults).toMatchSnapshot();
    });
    it("should count hidden fields", () => {
        expect(Object.values(reducedResults).filter(field => field.isHidden)).toHaveLength(1);
    });
    it("should count disabled fields", () => {
        expect(Object.values(reducedResults).filter(field => field.isDisabled)).toHaveLength(1);
    });
    it("should find overridden rule", () => {
        function isOverridden(c: RuleOverride.OverridableRuleContextInfo): boolean {
            return c.contextAttributeValue === "override";
        }
        const reducer = new FieldMetadataReducer(isOverridden);
        const res = reducer.reduce(entryPointResults);
        expect(res["3"].ruleResults[0].isOverridden).toBeTruthy();
    });
    it("should return results for one field with overridden values", () => {
        const reducer = new FieldMetadataReducer(p => p.rule.ruleName === "length-ov-critical");
        const fm = reducer.reduce(new EntryPointResult(FieldMetadataMocks.results));
        expect(Object.values(fm)).toHaveLength(1);
        const fmResults = Object.values(fm)[0].ruleResults;
        expect(fmResults).toHaveLength(4);
        const fmResultsByRuleName = fmResults.reduce(Reducer.toMap(x => x.ruleName), Reducer.Map());
        const overriddenFailed = fmResultsByRuleName.get("length-ov-critical");
        expect(overriddenFailed!.isFailed).toBeTruthy();
        expect(overriddenFailed!.isOverridable).toBeTruthy();
        expect(overriddenFailed!.isOverridden).toBeTruthy();
        const overriddenAssertFailed = fmResultsByRuleName.get("assert-ov-critical");
        expect(overriddenAssertFailed!.isFailed).toBeTruthy();
        expect(overriddenAssertFailed!.isOverridable).toBeTruthy();
        expect(overriddenAssertFailed!.isOverridden).toBeFalsy();
        expect(fm).toMatchSnapshot();
    });
    it("should reduce mandatory results", () => {
        const reduced = new FieldMetadataReducer()
            .reduce(new EntryPointResult(FieldMetadataMocks.mandatoryResults));
        expect(Object.values(reduced)).toHaveLength(1);
        const fmResults = Object.values(reduced)[0].ruleResults;
        expect(fmResults).toHaveLength(2);
        expect(fmResults).toMatchSnapshot();
    });
    it("should set undefined to isFailed field if validation was ignored", () => {
        const reducer = new FieldMetadataReducer();
        const reduced = reducer.reduce(new EntryPointResult({
            "Policy:1:state": {
                contextFieldInfo: {
                    contextId: "1",
                    contextName: "Policy",
                    fieldName: "state",
                    fieldPath: "state",
                    fieldType: "STRING"
                },
                ruleResults: [
                    {
                        kind: RuleEvaluationResults.Kind.VALIDATION,
                        conditionEvaluationResult: ConditionEvaluationResult.of(ConditionEvaluation.APPLICABLE),
                        overrideInfo: {
                            overridable: false,
                            overrideApplicable: false
                        },
                        payloadResult: payloadResultCreator.assertionFail({
                            kind: 2,
                            error: {
                                message: "test",
                                severity: "info"
                            }
                        }),
                        ruleInfo: {
                            context: "Policy",
                            payloadtype: Payloads.PayloadType.ASSERTION,
                            ruleName: "test",
                            targetPath: "state"
                        }
                    }
                ]
            }
        }));
        expect(reduced["Policy:1:state"].ruleResults[0].isFailed).toBeUndefined();
    });
    it("should set set mustBeEmpty failure as a rule result", () => {
        const reducer = new FieldMetadataReducer();
        const reduced = reducer.reduce(new EntryPointResult({
            "Policy:1:state": {
                contextFieldInfo: {
                    contextId: "1",
                    contextName: "Policy",
                    fieldName: "state",
                    fieldPath: "state",
                    fieldType: "STRING"
                },
                ruleResults: [
                    {
                        kind: RuleEvaluationResults.Kind.VALIDATION,
                        conditionEvaluationResult: ConditionEvaluationResult.of(ConditionEvaluation.APPLICABLE),
                        overrideInfo: {
                            overridable: false,
                            overrideApplicable: false
                        },
                        payloadResult: payloadResultCreator.usage(
                            PayloadBuilder.usage().is(Payloads.Validation.UsageType.mustBeEmpty),
                            false,
                            []
                        ),
                        ruleInfo: {
                            context: "Policy",
                            payloadtype: Payloads.PayloadType.USAGE,
                            ruleName: "test",
                            targetPath: "state"
                        }
                    }
                ]
            }
        }));
        const ruleResult = reduced["Policy:1:state"].ruleResults[0];
        expect(ruleResult.ruleName).toBe("test");
        expect(ruleResult.payloadResult.type).toBe(PayloadResultType.USAGE);
        expect(ruleResult.isFailed).toBe(true);
        expect(payloadResultTypeChecker.isEmpty(ruleResult.payloadResult)).toBe(true);
    });
});
