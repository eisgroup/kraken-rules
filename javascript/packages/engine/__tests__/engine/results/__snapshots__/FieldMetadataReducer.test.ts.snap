// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldMetadataReducer should contain 2 field metadatas 1`] = `
Object {
  "2": Object {
    "fieldType": "STRING",
    "id": "2",
    "info": Object {
      "getContextInstanceId": [Function],
      "getContextName": [Function],
    },
    "isDisabled": undefined,
    "isHidden": undefined,
    "resolvedTargetPath": "transactionDetails.txType",
    "ruleResults": Array [],
  },
  "3": Object {
    "fieldType": "STRING",
    "id": "3",
    "info": Object {
      "getContextInstanceId": [Function],
      "getContextName": [Function],
    },
    "isDisabled": undefined,
    "isHidden": undefined,
    "resolvedTargetPath": "accessTrackInfo.updatedBy",
    "ruleResults": Array [
      Object {
        "errorCode": "assert-generated",
        "errorMessage": "Value didn't match assertion: ",
        "isFailed": true,
        "isOverridable": true,
        "isOverridden": false,
        "payloadResult": Object {
          "message": Object {
            "errorCode": "assert-generated",
            "errorMessage": "Value didn't match assertion: ",
            "rawTemplateVariables": Array [],
            "templateVariables": Array [],
          },
          "success": false,
          "type": 1,
          "validationSeverity": "critical",
        },
        "rawTemplateVariables": Array [],
        "ruleName": "mock",
        "severity": "critical",
        "templateVariables": Array [],
      },
    ],
  },
  "4": Object {
    "fieldType": "INTEGER",
    "id": "4",
    "info": Object {
      "getContextInstanceId": [Function],
      "getContextName": [Function],
    },
    "isDisabled": undefined,
    "isHidden": undefined,
    "resolvedTargetPath": "termDetails.termNo",
    "ruleResults": Array [],
  },
  "present1": Object {
    "fieldType": "STRING",
    "id": "present1",
    "info": Object {
      "getContextInstanceId": [Function],
      "getContextName": [Function],
    },
    "isDisabled": true,
    "isHidden": undefined,
    "resolvedTargetPath": "state",
    "ruleResults": Array [],
  },
  "present2": Object {
    "fieldType": "STRING",
    "id": "present2",
    "info": Object {
      "getContextInstanceId": [Function],
      "getContextName": [Function],
    },
    "isDisabled": undefined,
    "isHidden": true,
    "resolvedTargetPath": "termDetails.termCd",
    "ruleResults": Array [],
  },
}
`;

exports[`FieldMetadataReducer should reduce mandatory results 1`] = `
Array [
  Object {
    "errorCode": "usage-generated",
    "errorMessage": "The field is mandatory",
    "isFailed": false,
    "isOverridable": false,
    "isOverridden": false,
    "payloadResult": Object {
      "message": Object {
        "errorCode": "usage-generated",
        "errorMessage": "The field is mandatory",
        "rawTemplateVariables": Array [],
        "templateVariables": Array [],
      },
      "success": true,
      "type": 64,
      "usageType": "mandatory",
      "validationSeverity": "critical",
    },
    "rawTemplateVariables": Array [],
    "ruleName": "mandatory-critical-true",
    "severity": "critical",
    "templateVariables": Array [],
  },
  Object {
    "errorCode": "usage-generated",
    "errorMessage": "The field is mandatory",
    "isFailed": true,
    "isOverridable": false,
    "isOverridden": false,
    "payloadResult": Object {
      "message": Object {
        "errorCode": "usage-generated",
        "errorMessage": "The field is mandatory",
        "rawTemplateVariables": Array [],
        "templateVariables": Array [],
      },
      "success": false,
      "type": 64,
      "usageType": "mandatory",
      "validationSeverity": "critical",
    },
    "rawTemplateVariables": Array [],
    "ruleName": "mandatory-critical-false",
    "severity": "critical",
    "templateVariables": Array [],
  },
]
`;

exports[`FieldMetadataReducer should return results for one field with overridden values 1`] = `
Object {
  "policy:0:state": Object {
    "fieldType": "STRING",
    "id": "policy:0:state",
    "info": Object {
      "getContextInstanceId": [Function],
      "getContextName": [Function],
    },
    "isDisabled": undefined,
    "isHidden": undefined,
    "resolvedTargetPath": "state",
    "ruleResults": Array [
      Object {
        "errorCode": "limit-generated",
        "errorMessage": "String contains more characters than 5",
        "isFailed": true,
        "isOverridable": true,
        "isOverridden": true,
        "payloadResult": Object {
          "length": 5,
          "message": Object {
            "errorCode": "limit-generated",
            "errorMessage": "String contains more characters than 5",
            "rawTemplateVariables": Array [],
            "templateVariables": Array [],
          },
          "success": false,
          "type": 16,
          "validationSeverity": "critical",
        },
        "rawTemplateVariables": Array [],
        "ruleName": "length-ov-critical",
        "severity": "critical",
        "templateVariables": Array [],
      },
      Object {
        "errorCode": "assert-generated",
        "errorMessage": "Value didn't match assertion: false",
        "isFailed": true,
        "isOverridable": true,
        "isOverridden": false,
        "payloadResult": Object {
          "message": Object {
            "errorCode": "assert-generated",
            "errorMessage": "Value didn't match assertion: false",
            "rawTemplateVariables": Array [],
            "templateVariables": Array [],
          },
          "success": false,
          "type": 1,
          "validationSeverity": "critical",
        },
        "rawTemplateVariables": Array [],
        "ruleName": "assert-ov-critical",
        "severity": "critical",
        "templateVariables": Array [],
      },
      Object {
        "errorCode": "assert-generated",
        "errorMessage": "Value didn't match assertion: false",
        "isFailed": true,
        "isOverridable": false,
        "isOverridden": false,
        "payloadResult": Object {
          "message": Object {
            "errorCode": "assert-generated",
            "errorMessage": "Value didn't match assertion: false",
            "rawTemplateVariables": Array [],
            "templateVariables": Array [],
          },
          "success": false,
          "type": 1,
          "validationSeverity": "critical",
        },
        "rawTemplateVariables": Array [],
        "ruleName": "assert-critical",
        "severity": "critical",
        "templateVariables": Array [],
      },
      Object {
        "errorCode": "limit-generated",
        "errorMessage": "String contains more characters than 5",
        "isFailed": true,
        "isOverridable": true,
        "isOverridden": false,
        "payloadResult": Object {
          "length": 5,
          "message": Object {
            "errorCode": "limit-generated",
            "errorMessage": "String contains more characters than 5",
            "rawTemplateVariables": Array [],
            "templateVariables": Array [],
          },
          "success": false,
          "type": 16,
          "validationSeverity": "warning",
        },
        "rawTemplateVariables": Array [],
        "ruleName": "length-ov-warning",
        "severity": "warning",
        "templateVariables": Array [],
      },
    ],
  },
}
`;
